### PR TITLE
refactor/hummingbot client performance

### PR DESCRIPTION
**Pull request by aarmoa**
**Before submitting this PR, please make sure**:

- [ ] Your code builds clean without any errors or warnings
- [ ] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Refactor to the module hummingbot.client.performance moving the functions inside the PerformanceMetric class. All references to those functions have been changed and now the class is used everywhere.
I have also made some changes in the functions implementation to remove redundant collection cycles, and extracted logic from long methods into smaller ones.
All changes where made after ensuring the module has a 99% coverage from unit tests. It should be safe to integrate the changes, but I would recommend to ask for QA review.

https://app.zenhub.com/workspaces/hummingbot-5f340755c5e17300167a8f30/issues/coinalpha/hummingbot/3383
https://github.com/CoinAlpha/hummingbot/issues/3383

### CODE REVIEW

**Review by keithbaum — APPROVED** [review]
LGTM

File: hummingbot/client/performance.py [hunk 1]
```diff
@@ -52,199 +54,234 @@ def __init__(self):
...
-    Calculates PnL, fees, Return % and etc...
-    :param exchange: the exchange or connector name
-    :param trading_pair: the trading market to get performance metrics
-    :param trades: the list of TradeFill or Trade object
-    :param current_balances: current user account balance
-    :return: A PerformanceMetrics object
-    """
-
+    @classmethod
+    async def create(cls, exchange: str, trading_pair: str, trades: List[Any], current_balances: Dict[str, Decimal]):
```
Comment (Nullably) [inline hummingbot/client/performance.py:58]: Can you add type hint for the returned value, please?

File: hummingbot/client/performance.py [hunk 2]
```diff
@@ -52,199 +54,234 @@ def __init__(self):
...
+        performance = PerformanceMetrics()
+        await performance._initialize_metrics(exchange, trading_pair, trades, current_balances)
+        return performance
+
+    @staticmethod
+    def position_order(open: list, close: list):
```
Comment (Nullably) [inline hummingbot/client/performance.py:64]: Missing type hint on returned value.

File: hummingbot/client/performance.py [hunk 3]
```diff
@@ -52,199 +54,234 @@ def __init__(self):
...
+                aggregated_amounts += order.amount
+            aggregated = group[0]
+            aggregated.price = aggregated_prices / len(group)
+            aggregated.amount = aggregated_amounts
+            aggregated_orders.append(aggregated)
+
+        return aggregated_orders
+
+    @staticmethod
+    def aggregate_position_order(buys: list, sells: list):
```
Comment (Nullably) [inline hummingbot/client/performance.py:107]: Missing type hint on returned value.

File: hummingbot/client/performance.py [hunk 4]
```diff
@@ -52,199 +54,234 @@ def __init__(self):
...
+        :param sells: a list of sell orders
+        :return: 2 lists containing aggregated amounts for buy and sell orders.
+        """
+        aggregated_buys = PerformanceMetrics.aggregate_orders(buys)
+        aggregated_sells = PerformanceMetrics.aggregate_orders(sells)
+
+        return aggregated_buys, aggregated_sells
+
+    @staticmethod
+    def derivative_pnl(long: list, short: list):
```
Comment (Nullably) [inline hummingbot/client/performance.py:120]: Missing type hint on returned value.

File: hummingbot/client/performance.py [hunk 5]
```diff
@@ -52,199 +54,234 @@ def __init__(self):
...
+
+                sht = self.position_order(sells_copy, buys_copy)
+                if sht is not None:
+                    short.append(sht)
+                if lng is None and sht is None:
+                    break
+
+            self.trade_pnl = Decimal(str(sum(self.derivative_pnl(long, short))))
+
+    async def _initialize_metrics(self,
```
Comment (Nullably) [inline hummingbot/client/performance.py:245]: Missing type hint on returned value.

**Review by Nullably — CHANGES_REQUESTED** [review]
Minor change requested for type hints, the rest looks good to me.

File: hummingbot/client/performance.py (hunk 5 above)
Comment (aarmoa) [reply hummingbot/client/performance.py:245]: This method initializes the internal status of the object, but it does not return anything.

**Comment by aarmoa** [comment]
I have commited the changes requested by @Nullably

**Review by dennisocana — APPROVED** [review]

**Comment by aarmoa** [comment]
This PR was merged by mistake in refactor/unit_tests
Since it includes model changes, it has to be reviewed by QA.
The merge was reverted and I have created a new PR to merge the changes: https://github.com/CoinAlpha/hummingbot/pull/3395

**Pull request merged**